keps/sig-storage/20200127-recover-from-resize.md

```diff
@@ -0,0 +1,153 @@
+
+
+
+### Goals
+
+- Allow users to cancel previously issued volume expansion requests, assuming they are not yet successful or have failed.
+- Allow users to retry expansion request with smaller value than original requested size in `pvc.Spec.Resources`, assuming previous requests are not yet successful or have failed.
+
+### Non-Goals
+
+- As part of this KEP we do not intend to actually support shrinking of underlying volume.
+
+## Proposal
+
+As part of this proposal, we are mainly proposing three changes:
+
+1. Relax API validation on PVC update so as reducing `pvc.Spec.Resoures` is allowed, as long requested size is `>=pvc.Status.Capacity`.
+2. Add a new field in PVC api called - `pvc.Spec.AllocatedResources` for purpose of tracking used storage size. This field is only allowed to increase.
+3. Update quota code to use `max(pvc.Spec.Resources, pvc.Spec.AllocatedResources)` when evaluating usage for PVC.
+
+### Implementation
+
+We propose that by relaxing validation on PVC update to allow users to reduce `pvc.Spec.Resources`, it becomes possible to cancel previously issued expansion requests or retry expansion with a lower value if previous request has not been attemptd. In general - we know that volume plugins are designed to never peform actual shrinking of the volume, for both in-tree and CSI volumes. Moreover if a previously issued expansion has been successful and user
+reduces the PVC request size, for both CSI and in-tree plugins they are designed to return a successful response with NO-OP. So, reducing requested size will be a safe operation and will never result in data loss or actual shrinking of volume.
+
+We however do have a problem with quota calculation because if a previously issued expansion is successful but is not recorded(or partially recorded) in api-server and user reduces requested size of the PVC, then quota controller will assume it as actual shrinking of volume and reduce used storage size by the user(incorrectly). Since we know actual size of the volume only after performing expansion(either on node or controller), allowing quota to be reduced on PVC size reduction will allow an user to abuse the quota system.
+
+To solve aformentioned problem - we propose that, a new field will be added to PVC, called `pvc.Spec.AllocatedResources`. This field is only allowed to increase and will be set by the api-server to value in `pvc.Spec.Resources` as long as `pvc.Spec.Resources > pvc.Spec.AllocatedResources`.  Quota controller code will be updated to use `max(pvc.Spec.Resources, pvc.Spec.AllocatedResources)` when calculating usage of the PVC under questions. This does mean that - if user expanded PVC to a size that it failed to expand to and user retries expansion with lower value, the used quota will still use value reported by failed expansion request.
```

Comment:
I think you mean: This field is only allowed to increase and will be set by the api-server to value in `pv.Spec.Capacity` not `pvc.Spec.Resources`. Otherwise this is useless. The former is the actual size of the volume, the latter is the size the user controls.